### PR TITLE
#8 - Admin Role Approval and Admin Assign Doctors to Patient

### PR DESCRIPTION
**Pull request by DrakeHunterK7**
# General info

**Issue number**: #8

**Task description**: 

 - Admin can now approve/reject role requests when users sign up to CoviCare
 - Admin can now assign doctors to patients via the Patient List tab/page.
 
# Testing

**Test coverage**:

### frontend

- manual test all features: all passed
- Some basic tests for PatientBox and PatientList component implemented. More to come shortly. 

# Additional notes

**Note to reviewers**:

- Sign-up and create a new patient account so a role request can be sent to the admin
- Then log in using the following credentials to the admin account => username: admin@gmail.com, password: admin1234&
- Now you will see your new user as a role request on the Role Requests page
- Accept your new user
- Now go to the Patients List page, you'll see your new user in the list of patients. 
- Click on your patient box, and now click on Assign Doctor
- You will see a list of doctors fetched from the database. Assign any doctor to the patient. 
- Now follow the same steps to assign another doctor to the patient
- Notice the patient count of both doctors as you make the switch
- Basic features are now tested. You can play around now and see if there are any bugs.

### CODE REVIEW

File: frontend/src/components/AdminComponents/PatientBox.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+import { Button } from "react-bootstrap";
+import AccordionBody from "react-bootstrap/esm/AccordionBody";
+import { Modal } from "react-bootstrap";
+import { ListGroup } from "react-bootstrap";
+
+import axios from "axios";
+
+const eventKey = "0";
+const patient = null;
+const currentUser = null;
```
Comment (AC159) [inline frontend/src/components/AdminComponents/PatientBox.js:15]: Please remove these 3 values as they are not used anywhere in the code

File: frontend/src/components/AdminComponents/PatientBox.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  const handleClose = () => setShow(false);
+  const handleShow = () => setShow(true);
+
+  const openDoctorList = () => {
+    handleShow();
+    fetchDoctorList();
+  }
+
+  const selectDoctor = (id, name) => {
+    selectedDoctorID=id;
```
Comment (AC159) [inline frontend/src/components/AdminComponents/PatientBox.js:37]: When selecting a doctor in the UI, it is not clear if the doctor was selected or not. Maybe add check boxes to make it clear?

File: frontend/src/components/AdminComponents/PatientBox.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+          </Button>
+          </div>
+          
+        </ListGroup.Item> )}
+    </ListGroup>;
+  } 
+
+  useEffect(()=> {
+    
+  }, []);
```
Comment (AC159) [inline frontend/src/components/AdminComponents/PatientBox.js:94]: remove this empty function

File: frontend/src/components/AdminComponents/PatientBox.module.css [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  width: 100%;
+}
+
+.card {
+  width: 100%;
+  margin-bottom: 10px;
+}
+
+.doctor-select-button {
+
```
Comment (AC159) [inline frontend/src/components/AdminComponents/PatientBox.module.css:11]: I think this is the missing style for the "select doctor" button

File: frontend/src/components/AdminComponents/PatientBox.test.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
+import {render, screen} from '@testing-library/react';
+import {BrowserRouter, MemoryRouter} from 'react-router-dom';
+import userEvent from '@testing-library/user-event'
+
+import PatientBox from './PatientBox';
+import fetchDoctorList from './PatientBox';
```
Comment (AC159) [inline frontend/src/components/AdminComponents/PatientBox.test.js:6]: remove unused imports

File: frontend/src/components/AdminComponents/RoleRequestBox.js [hunk 6]
```diff
@@ -1,26 +1,70 @@
...
+    })
+    .catch((error) => {
+      console.log(error.response);
+    })
+
+    setDisplayRB(false);
+  }
+
+  const rejectUser = () => {
+    const approveInfo = {
```
Comment (AC159) [inline frontend/src/components/AdminComponents/RoleRequestBox.js:35]: This variable needs to be renamed to "rejectInfo"

File: frontend/src/components/Authentication/login/Login.js [hunk 7]
```diff
@@ -17,6 +17,7 @@ function Login(props) {
       console.log("User: ", user);
       // get the requested route from local storage
       const redirectRoute = sessionStorage.getItem("requestedRoute");
+      console.log("Here be the route:");
```
Comment (AC159) [inline frontend/src/components/Authentication/login/Login.js:20]: Unnecessary console log

**Comment by DrakeHunterK7** [comment]
Thank you for the very helpful review, @AC159 . I will fix all these shortly.

File: frontend/src/components/Tabs/AdminTabs/PatientList.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+function PatientList(props) {
+
+  let {currentUser} = useAuth();
+
+  const [patientList, setPatientList] = useState([]);
+ 
+
+  useEffect(() => {
+    console.log("Patient List Tab is Loaded!");
+    fetchListOfPatients();
```
Comment (AC159) [inline frontend/src/components/Tabs/AdminTabs/PatientList.js:20]: We can put this function inline inside the useEffect hook

File: frontend/src/components/Tabs/AdminTabs/PatientList.module.css [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    padding-bottom: 0px;
+    font-weight: 700;
+  }
+  
+  .request-container {
+    padding-bottom: 10px;
+  }
+
+  .patient-box{
+
```
Comment (AC159) [inline frontend/src/components/Tabs/AdminTabs/PatientList.module.css:17]: remove empty css selectors

File: frontend/src/components/Tabs/AdminTabs/PatientList.test.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import {render, screen} from '@testing-library/react';
+import {BrowserRouter, MemoryRouter} from 'react-router-dom';
+import userEvent from '@testing-library/user-event'
```
Comment (AC159) [inline frontend/src/components/Tabs/AdminTabs/PatientList.test.js:3]: remove unused imports

File: frontend/src/components/Tabs/AdminTabs/PatientList.test.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+import PatientList from "./PatientList.js"
+import {AuthContext} from '../../Authentication/FirebaseAuth/FirebaseAuth';
+
+import axios from 'axios'
+jest.mock('axios');
+
+describe('visual test of the PatientList qcomponent', () => {
```
Comment (AC159) [inline frontend/src/components/Tabs/AdminTabs/PatientList.test.js:11]: fix typo "qcomponent"

File: frontend/src/components/Tabs/AdminTabs/RoleRequest.js [hunk 12]
```diff
@@ -1,31 +1,159 @@
...
+import { useAuth } from "../../Authentication/FirebaseAuth/FirebaseAuth";
+import axios from "axios";
 
 function RoleRequest(props) {
 
+  let {currentUser} = useAuth();
+
+  const [patientList, setPatientList] = useState([]);
+  const [doctorList, setDoctorList] = useState([]);
+  const [adminList, setAdminList] = useState([]);
```
Comment (AC159) [inline frontend/src/components/Tabs/AdminTabs/RoleRequest.js:14]: remove unused state variable

File: frontend/src/components/Tabs/AdminTabs/RoleRequest.js [hunk 13]
```diff
@@ -1,31 +1,159 @@
...
+  </div>;  
+    }
+    
+  }
+
+  /*return (
+    <div className={styles["role-outer-container"]}>
+      <div className={styles["todays-new-title"]}>Patient List</div>
+      <hr />
+      <div>
```
Comment (AC159) [inline frontend/src/components/Tabs/AdminTabs/RoleRequest.js:120]: remove commented code

File: repository/AdminRepository.js [hunk 14]
```diff
@@ -23,12 +23,79 @@ class AdminRepository {
...
   async rejectUser(userId) {
     return await this.mongo.db('test').collection('user').updateOne({uid: userId}, {$set: {userStatus: 'REJECTED'}});
   }
+
+  async fetchPatients() {
+    return await this.mongo.db('test').collection('user')
+        .find({
+          userType: 'patient',
+          userStatus: 'APPROVED'})
+    // patientInfo: {doctor: null}})
```
Comment (AC159) [inline repository/AdminRepository.js:42]: remove this commented line

**Review by AC159 — CHANGES_REQUESTED** [review]
See comments

**Pull request merged**